src/formats/pnm.c

```diff
@@ -0,0 +1,287 @@
+static int decode_plain(struct image_frame* f, struct pnm_iter* it, int type,
+                        int maxval)
+{
+    for (size_t y = 0; y < f->height; ++y) {
+        argb_t* dst = f->data + y * f->width;
+        for (size_t x = 0; x < f->width; ++x) {
+            argb_t pix = ARGB_SET_A(0xff);
+            if (type == PNM_PBM) {
+                int bit = pnm_readint(it);
+                if (bit < 0)
+                    return bit;
+                if (bit > 1)
+                    return PNM_EOVF;
+                pix |= bit - 1;
+            } else if (type == PNM_PGM) {
+                int v = pnm_readint(it);
+                if (v < 0)
+                    return v;
+                if (v > maxval)
+                    return PNM_EOVF;
+                if (maxval != UINT8_MAX)
+                    v = div_near(v * UINT8_MAX, maxval);
+                pix |= ARGB_SET_R(v) | ARGB_SET_G(v) | ARGB_SET_B(v);
+            } else {
+                int r = pnm_readint(it);
+                if (r < 0)
+                    return r;
+                if (r > maxval)
+                    return PNM_EOVF;
+                int g = pnm_readint(it);
+                if (g < 0)
+                    return g;
+                if (g > maxval)
+                    return PNM_EOVF;
+                int b = pnm_readint(it);
+                if (b < 0)
+                    return b;
+                if (b > maxval)
+                    return PNM_EOVF;
+                if (maxval != UINT8_MAX) {
+                    r = div_near(r * UINT8_MAX, maxval);
+                    g = div_near(g * UINT8_MAX, maxval);
+                    b = div_near(b * UINT8_MAX, maxval);
+                }
+                pix |= ARGB_SET_R(r) | ARGB_SET_G(g) | ARGB_SET_B(b);
+            }
+            dst[x] = pix;
+        }
+    }
+    return 0;
+}
+
+// Decode a raw/binary PNM file; returns 0 on success, an error code otherwise
+static int decode_raw(struct image_frame* f, struct pnm_iter* it, int type,
+                      int maxval)
+{
+    // PGM and PPM use bpc (bytes per channel) bytes for each channel depending
+    // on the max, with 1 channel for PGM and 3 for PPM; PBM pads each row to
+    // the nearest whole byte
+    size_t bpc = maxval <= UINT8_MAX ? 1 : 2;
+    size_t rowsz = type == PNM_PBM ? div_ceil(f->width, 8)
+                                   : f->width * bpc * (type == PNM_PGM ? 1 : 3);
+    if (it->end < it->pos + f->height * rowsz)
+        return PNM_EEOF;
+    for (size_t y = 0; y < f->height; ++y) {
+        argb_t* dst = f->data + y * f->width;
+        const uint8_t* src = it->pos + y * rowsz;
+        for (size_t x = 0; x < f->width; ++x) {
+            argb_t pix = ARGB_SET_A(0xff);
+            if (type == PNM_PBM) {
+                int bit = (src[x / 8] >> (7 - x % 8)) & 1;
+                pix |= bit - 1;
+            } else if (type == PNM_PGM) {
+                int v;
+                if (bpc == 1)
```

Comment:
Please, add `{}` for `if` when you have more then 2 lines in a block (include else). I prefer squares even for one-lined if's.